internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,396 @@
+		eliInfo.BasicLimitInformation.LimitFlags = basicLimitFlags
+		_, err := windows.SetInformationJobObject(job.jobHandle, windows.JobObjectExtendedLimitInformation, uintptr(unsafe.Pointer(&eliInfo)), uint32(unsafe.Sizeof(eliInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set extended limit info on job object: %s", err)
+		}
+	}
+
+	if limits.CPULimit != 0 || limits.CPUWeight != 0 {
+		cpuInfo := winapi.JOBOBJECT_CPU_RATE_CONTROL_INFORMATION{}
+		if limits.CPUWeight != 0 {
+			if limits.CPUWeight < ProcessorWeightMin || limits.CPUWeight > ProcessorWeightMax {
+				return fmt.Errorf("processor weight value of `%d` is invalid", limits.CPUWeight)
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_WEIGHT_BASED
+			cpuInfo.Value = limits.CPUWeight
+		} else if limits.CPULimit != 0 {
+			if limits.CPULimit < ProcessorLimitMin || limits.CPULimit > ProcessorLimitMax {
+				return fmt.Errorf("processor rate of `%d` is invalid", limits.CPULimit)
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP
+			cpuInfo.Value = limits.CPULimit
+		}
+		_, err := windows.SetInformationJobObject(job.jobHandle, windows.JobObjectCpuRateControlInformation, uintptr(unsafe.Pointer(&cpuInfo)), uint32(unsafe.Sizeof(cpuInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set cpu limit info on job object: %s", err)
+		}
+	}
+
+	if limits.MaxBandwidth != 0 || limits.MaxIops != 0 {
+		ioInfo := winapi.JOBOBJECT_IO_RATE_CONTROL_INFORMATION{
+			ControlFlags: winapi.JOB_OBJECT_IO_RATE_CONTROL_ENABLE,
+		}
+		if limits.MaxBandwidth != 0 {
+			ioInfo.MaxBandwidth = limits.MaxBandwidth
+		}
+		if limits.MaxIops != 0 {
+			ioInfo.MaxIops = limits.MaxIops
+		}
+		_, err := winapi.SetIoRateControlInformationJobObject(job.jobHandle, &ioInfo)
+		if err != nil {
+			return fmt.Errorf("failed to set IO limit info on job object: %s", err)
+		}
+	}
+	return nil
+}
+
+// Create creates a job object, attaches an IO completion port to use
+// for notifications and then returns an object with the corresponding handles.
+func Create(ctx context.Context, name string) (*JobObject, error) {
+	// If the IOCP we'll be using to receive messages for all jobs hasn't been
+	// created, create it and start polling.
+	ioInitOnce.Do(func() {
+		h, err := windows.CreateIoCompletionPort(windows.InvalidHandle, 0, 0, 0xffffffff)
+		if err != nil {
+			initIOErr = err
+			return
+		}
+		ioCompletionPort = h
+		go pollIOCP(ctx, h)
+	})
+
+	if initIOErr != nil {
+		return nil, initIOErr
+	}
+
+	var (
+		jobName *uint16
+		err     error
+	)
+	if name != "" {
+		jobName, err = windows.UTF16PtrFromString(name)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	jobHandle, err := windows.CreateJobObject(nil, jobName)
+	if err != nil {
+		return nil, err
+	}
+
+	iocpChan := make(chan interface{}, 10)
+	jobMap[uintptr(jobHandle)] = iocpChan
+
+	defer func() {
+		if err != nil {
```

Comment:
When we do this kind of error check inside a defer, we generally want to also declare `err` as a named return value. Otherwise it's pretty easy to get in a state where `err` is redeclared inside a smaller scope, which can break the defer's check.